src/peft/tuners/lora/bnb.py

```diff
@@ -62,39 +51,55 @@ def merge(self):
                     f"Already following adapters were merged {','.join(self.merged_adapters)}. "
                     f"You are now additionally merging {','.join(self.active_adapters)}."
                 )
+
+            weight = self.base_layer.weight
+            state = self.base_layer.state
+            if state.SCB is None:
+                state.SCB = weight.SCB
+
+            # Dequantize the result of identity matrix and int8 weight because bitsandbytes does not support int8
+            # dequantization directly
+            im = torch.eye(weight.data.shape[-1]).contiguous().half().to(weight.device)
+            im, imt, SCim, SCimt, coo_tensorim = bnb.functional.double_quant(im)
+            im, Sim = bnb.functional.transform(im, "col32")
+            if state.CxB is None:
+                state.CxB, state.SB = bnb.functional.transform(weight.data, to_order=state.formatB)
+            out32, Sout32 = bnb.functional.igemmlt(im, state.CxB, Sim, state.SB)
+            output = bnb.functional.mm_dequant(out32, Sout32, SCim, state.SCB, bias=None).t()
+
             for active_adapter in self.active_adapters:
                 if active_adapter not in self.lora_A.keys():
                     continue
                 warnings.warn(
                     "Merge lora module to 8-bit linear may get different generations due to rounding errors."
                 )
                 lora_data = self.get_delta_weight(active_adapter)
-
-                if self.state.SCB is None:
-                    self.state.SCB = self.weight.SCB
-                # Dequantize the result of identity matrix and int8 weight because bitsandbytes does not support int8
-                # dequantization directly
-                im = torch.eye(self.weight.data.shape[-1]).contiguous().half().to(self.weight.device)
-                im, imt, SCim, SCimt, coo_tensorim = bnb.functional.double_quant(im)
-                im, Sim = bnb.functional.transform(im, "col32")
-
-                if self.state.CxB is None:
-                    self.state.CxB, self.state.SB = bnb.functional.transform(
-                        self.weight.data, to_order=self.state.formatB
-                    )
-                out32, Sout32 = bnb.functional.igemmlt(im, self.state.CxB, Sim, self.state.SB)
-                output = bnb.functional.mm_dequant(out32, Sout32, SCim, self.state.SCB, bias=None).t()
                 w_data = output.to(lora_data.dtype).to(lora_data.device) + lora_data
-                self.weight = bnb.nn.Int8Params(
-                    w_data.to("cpu"), requires_grad=False, has_fp16_weights=self.weight.has_fp16_weights
-                ).to(self.weight.device)
-                self.state.reset_grads()
+                self.base_layer.weight = bnb.nn.Int8Params(
```

Comment:
I believe that on line 77, the output needs to be the newer base weights but they are the initial ones itself leading to this being incorrect when multiple adapters are active.

Reply:
To clarify, do you mean that the `output`, as computed in line 68, should instead be computed inside the loop, i.e. it changes on each iteration? Would that also mean that all the other lines above that, like computing `out32` and `Sout32`, need to be computed inside the loop?

To me, it looked like these values wouldn't change on each loop, but maybe I'm missing some subtle dependency.

Reply:
Done